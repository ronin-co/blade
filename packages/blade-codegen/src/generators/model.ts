import { SyntaxKind, factory } from 'typescript';

import { identifiers, typeArgumentIdentifiers } from '@/src/constants/identifiers';
import { DEFAULT_FIELD_SLUGS } from '@/src/constants/schema';
import { triggerOptionsInterface } from '@/src/declarations';
import { convertToPascalCase } from '@/src/utils/slug';
import { mapRoninFieldToTypeNode, remapNestedFields } from '@/src/utils/types';

import type { TypeAliasDeclaration, TypeParameterDeclaration } from 'typescript';

import type { ModelField } from '@/src/types/model';
import type { PopulatedModel } from 'blade-compiler';

/**
 * Generate all required type definitions for a provided schema model.
 *
 * The plural model type will be mapped to an array of the singular model type and
 * extended with the plural model properties.
 *
 * @param models - All models of the addressed space.
 *
 * @returns An array of type nodes to be added to the `index.d.ts` file.
 */
export const generateModelTypes = (
  models: Array<PopulatedModel>,
): Array<TypeAliasDeclaration> => {
  const nodes = new Array<TypeAliasDeclaration>(triggerOptionsInterface);

  for (const model of models) {
    const fields: Array<ModelField> = Object.entries(model.fields)
      .map(([slug, field]) => ({ ...field, slug }) as ModelField)
      .filter((field) => !DEFAULT_FIELD_SLUGS.includes(field.slug));

    const modelIdentifier = {
      singular: factory.createIdentifier(convertToPascalCase(model.slug)),
      plural: factory.createIdentifier(convertToPascalCase(model.pluralSlug)),
    };

    const hasLinkFields = fields.some(
      (field) =>
        field.type === 'link' && models.some((model) => model.slug === field.target),
    );

    const modelInterfaceTypeParameters = new Array<TypeParameterDeclaration>();
    const linkFieldKeys = fields
      .filter((field) => field.type === 'link')
      .map((field) =>
        factory.createLiteralTypeNode(factory.createStringLiteral(field.slug)),
      );

    /**
     * ```ts
     * <U extends Array<...> | 'all' = []>
     * ```
     */
    const usingGenericDec = factory.createTypeParameterDeclaration(
      undefined,
      typeArgumentIdentifiers.using,
      factory.createUnionTypeNode([
        factory.createTypeReferenceNode(identifiers.primitive.array, [
          factory.createUnionTypeNode(linkFieldKeys),
        ]),
        factory.createLiteralTypeNode(factory.createStringLiteral('all')),
      ]),
      factory.createTupleTypeNode([]),
    );

    if (hasLinkFields) modelInterfaceTypeParameters.push(usingGenericDec);

    /**
     * ```ts
     * SchemaSlugSchema<U>
     * ```
     */
    const modelSchemaName = factory.createTypeReferenceNode(
      modelIdentifier.singular,
      hasLinkFields
        ? [factory.createTypeReferenceNode(typeArgumentIdentifiers.using)]
        : [],
    );

    /**
     * ```ts
     * export type SchemaSlug<U extends Array<...> | 'all' = []> = ResultRecord & {
     *  // ...
     * };
     * ```
     */
    const singularModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      modelIdentifier.singular,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        factory.createTypeReferenceNode(
          factory.createQualifiedName(
            identifiers.namespace.utils.name,
            identifiers.namespace.utils.resultRecord,
          ),
        ),
        factory.createTypeLiteralNode(
          remapNestedFields(fields)
            .sort(([a], [b]) => a.localeCompare(b))
            .map(([fieldSlug, field]) => {
              if (Array.isArray(field)) {
                const sortedFields = field.sort((a, b) => a.slug.localeCompare(b.slug));
                return factory.createPropertySignature(
                  undefined,
                  fieldSlug,
                  undefined,
                  factory.createTypeLiteralNode(
                    sortedFields.map((nestedField) =>
                      factory.createPropertySignature(
                        undefined,
                        nestedField.slug,
                        undefined,
                        factory.createUnionTypeNode(
                          mapRoninFieldToTypeNode(nestedField, models),
                        ),
                      ),
                    ),
                  ),
                );
              }

              return factory.createPropertySignature(
                undefined,
                fieldSlug,
                undefined,
                factory.createUnionTypeNode(mapRoninFieldToTypeNode(field, models)),
              );
            }),
        ),
      ]),
    );

    /**
     * ```ts
     * Array<SchemaSlug>;
     * ```
     */
    const pluralModelArrayTypeDec = factory.createTypeReferenceNode(
      identifiers.primitive.array,
      [modelSchemaName],
    );

    /**
     * ```ts
     * {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelPaginationPropsTypeDec = factory.createTypeLiteralNode([
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreBefore'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreAfter'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
    ]);

    /**
     * ```ts
     * export type SchemaPluralSlug<U extends Array<...> | 'all' = []> = Array<SchemaSlug> & {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      modelIdentifier.plural,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        pluralModelArrayTypeDec,
        pluralModelPaginationPropsTypeDec,
      ]),
    );

    nodes.push(singularModelTypeDec, pluralModelTypeDec);
  }

  return nodes;
};
