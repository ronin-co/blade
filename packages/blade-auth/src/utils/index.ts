import { EmptyFieldsError, InvalidPermissionsError } from 'blade/server/errors';
import { type JWTPayload, verifyJWT } from 'blade/server/utils';
import type { SetQueryInstructions, TriggerOptions } from 'blade/types';

/**
 * Check given object for empty values. Throws an error if the object contains
 * empty values.
 *
 * @param to Object to check for empty values.
 * @param allowed Optional array of keys to exclude. Used in cases where `null`
 * is allowed as a value.
 */
export const avoidEmptyFields = (
  to: SetQueryInstructions['to'],
  allowed?: Array<string>,
) => {
  // We'll no longer have to use `(to || {})` once the `to` property on
  // `SetQueryInstructions` is guaranteed to be defined, which it already
  // should be.
  const emptyFields = Object.keys(to || {}).filter(
    (key) =>
      !(allowed || []).includes(key) && typeof to?.[key] !== 'boolean' && !to?.[key],
  );

  if (emptyFields.length > 0) {
    throw new EmptyFieldsError({ fields: emptyFields });
  }
};

/**
 * Obtains the secret used to sign authentication tokens.
 *
 * @returns The secret.
 */
export const getAuthSecret = (): string => {
  let secret = import.meta.env.BLADE_AUTH_SECRET;

  if (!secret) {
    if (import.meta.env.BLADE_ENV === 'development') {
      secret = 'default-secret-1234';
    } else {
      throw new Error('Please add a `BLADE_AUTH_SECRET` environment variable.');
    }
  }

  return secret;
};

/**
 * Retrieve the account that authored the incoming query.
 *
 * @param triggerOptions - The options object passed to the surrounding effect.
 *
 * @returns The ID of the account that authored the incoming query.
 */
export const getSessionCookie = async (
  triggerOptions: TriggerOptions,
): Promise<JWTPayload> => {
  const token = triggerOptions.cookies.session;

  let sessionId: string | null = null;
  let accountId: string | null = null;

  if (token) {
    try {
      const tokenPayload = await verifyJWT(token, getAuthSecret());

      sessionId = (tokenPayload?.sub as string) || null;
      accountId = (tokenPayload?.aud as string) || null;
    } catch (err) {
      console.warn(`Failed to decode token: \`${(err as Error).message}\``);
    }
  }

  if (!(sessionId && accountId)) {
    throw new InvalidPermissionsError({
      message: 'No session provided for authentication.',
    });
  }

  return { sessionId, accountId };
};

/**
 * Generate pseudo-random unique identifiers.
 *
 * @returns A unique identifier.
 */
export const generateUniqueId = (): string => {
  return crypto.getRandomValues(new Uint32Array(1))[0].toString();
};

export const EMAIL_VERIFICATION_COOLDOWN = 20_000; // 20s
