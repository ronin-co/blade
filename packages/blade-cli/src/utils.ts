export * from '@/src/utils/session';
