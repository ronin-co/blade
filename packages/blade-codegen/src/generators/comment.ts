import { READABLE_DML_QUERY_TYPES } from '@/src/constants/schema';

import type { QueryType } from '@/src/types/query';
import type { PopulatedModel } from 'blade-compiler';

interface GenerateQueryTypeCommentResult {
  singular: string;
  plural: string;
}

/**
 * Generate a text comment for a provided model & a given query type.
 *
 * @param modelName - The name of the model to generate a comment for.
 * @param queryType - The query type to generate a comment for.
 *
 * @returns An object containing both the singular and plural comment strings.
 */
export const generateQueryTypeComment = (
  model: PopulatedModel,
  queryType: QueryType,
): GenerateQueryTypeCommentResult => ({
  singular: `* ${READABLE_DML_QUERY_TYPES[queryType]} a single ${model.name ?? model.slug} record `,
  plural: `* ${READABLE_DML_QUERY_TYPES[queryType]} multiple ${model.name ?? model.slug} records `,
});
