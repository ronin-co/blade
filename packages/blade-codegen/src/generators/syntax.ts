import { factory } from 'typescript';

import { identifiers, typeArgumentIdentifiers } from '@/src/constants/identifiers';
import { DEFAULT_FIELD_SLUGS } from '@/src/constants/schema';
import { sharedQueryOptionsParameter } from '@/src/declarations';
import { convertToPascalCase } from '@/src/utils/slug';

import type {
  Identifier,
  IntersectionTypeNode,
  TypeAliasDeclaration,
  TypeNode,
} from 'typescript';

import type { Model } from '@/src/types/model';

/**
 * @todo(@nurodev): Add documentation
 */
export const generateUsingSyntax = (
  model: Model,
  modelNode: TypeNode,
  promise: boolean,
  plural: boolean,
): IntersectionTypeNode => {
  const hasLinkFields = Object.values(model.fields).some(
    (field) => field.type === 'link',
  );
  if (!hasLinkFields)
    return factory.createIntersectionTypeNode([
      factory.createExpressionWithTypeArguments(
        identifiers.blade.reducedFunction,
        undefined,
      ),
      factory.createFunctionTypeNode(
        [
          factory.createTypeParameterDeclaration(
            undefined,
            typeArgumentIdentifiers.default,
            undefined,
            modelNode,
          ),
        ],
        [
          factory.createParameterDeclaration(
            undefined,
            undefined,
            'value',
            undefined,
            factory.createIndexedAccessTypeNode(
              factory.createTypeReferenceNode(identifiers.compiler.combinedInstructions),
              factory.createLiteralTypeNode(factory.createStringLiteral('using')),
            ),
          ),
        ],
        promise
          ? factory.createTypeReferenceNode(identifiers.primitive.promise, [
              factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
            ])
          : factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
      ),
    ]);

  /**
   * @todo(@nurodev): Add documentation
   */
  const arrayFieldsType = factory.createUnionTypeNode([
    factory.createTypeReferenceNode(identifiers.primitive.array, [
      factory.createUnionTypeNode(
        Object.entries(model.fields)
          .filter(([, field]) => field.type === 'link')
          .map(([name]) =>
            factory.createLiteralTypeNode(factory.createStringLiteral(name)),
          ),
      ),
    ]),
    factory.createLiteralTypeNode(factory.createStringLiteral('all')),
  ]);

  /**
   * @todo(@nurodev): Add documentation
   */
  const baseModelWithFields = factory.createTypeReferenceNode(
    factory.createIdentifier(convertToPascalCase(model.slug)),
    [factory.createTypeReferenceNode(typeArgumentIdentifiers.using)],
  );
  const modelNodeWithFields = promise
    ? factory.createTypeReferenceNode(identifiers.primitive.promise, [
        baseModelWithFields,
      ])
    : baseModelWithFields;

  /**
   * @todo(@nurodev): Add documentation
   */
  const inferredCallSignature = factory.createCallSignature(
    [
      factory.createTypeParameterDeclaration(
        undefined,
        typeArgumentIdentifiers.using,
        arrayFieldsType,
      ),
    ],
    [
      factory.createParameterDeclaration(
        undefined,
        undefined,
        'fields',
        undefined,
        factory.createTypeReferenceNode(typeArgumentIdentifiers.using),
      ),
    ],
    plural
      ? modelNodeWithFields
      : factory.createUnionTypeNode([
          modelNodeWithFields,
          factory.createLiteralTypeNode(factory.createNull()),
        ]),
  );

  /**
   * @todo(@nurodev): Add documentation
   */
  const overrideCallSignaure = factory.createCallSignature(
    [
      factory.createTypeParameterDeclaration(
        undefined,
        typeArgumentIdentifiers.default,
        undefined,
        modelNode,
      ),
    ],
    [
      factory.createParameterDeclaration(
        undefined,
        undefined,
        'fields',
        undefined,
        arrayFieldsType,
      ),
    ],
    promise
      ? factory.createTypeReferenceNode(identifiers.primitive.promise, [
          factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
        ])
      : factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
  );

  return factory.createIntersectionTypeNode([
    factory.createExpressionWithTypeArguments(
      identifiers.blade.reducedFunction,
      undefined,
    ),
    factory.createTypeLiteralNode([inferredCallSignature, overrideCallSignaure]),
  ]);
};

/**
 * @todo(@nurodev): Add documentation
 */
export const generateWithSyntax = (
  name: string | Identifier,
  modelNode: TypeNode,
  model: Model,
  promise: boolean,
): TypeAliasDeclaration => {
  const modelUserFieldEntries = Object.entries(model.fields).filter(
    ([slug]) => !DEFAULT_FIELD_SLUGS.some((field) => field.includes(slug)),
  );

  return factory.createTypeAliasDeclaration(
    undefined,
    name,
    undefined,
    factory.createIntersectionTypeNode([
      factory.createTypeReferenceNode(
        factory.createQualifiedName(identifiers.namespace.utils.name, name),
        [modelNode],
      ),
      factory.createTypeLiteralNode(
        modelUserFieldEntries.map(([slug, field]) => {
          /**
           * @todo(@nurodev): Add documentation
           */
          const baseFieldProperty = factory.createIndexedAccessTypeNode(
            factory.createTypeReferenceNode(convertToPascalCase(model.slug)),
            factory.createLiteralTypeNode(factory.createStringLiteral(slug)),
          );

          const parameterNodes = new Array<TypeNode>(baseFieldProperty);
          if (field.type === 'link') {
            /**
             * @todo(@nurodev): Add documentation
             */
            const nestedFieldProperty = factory.createTypeReferenceNode(
              identifiers.primitive.partial,
              [
                factory.createIndexedAccessTypeNode(
                  factory.createTypeReferenceNode(convertToPascalCase(model.slug), [
                    factory.createTupleTypeNode([
                      factory.createLiteralTypeNode(factory.createStringLiteral(slug)),
                    ]),
                  ]),
                  factory.createLiteralTypeNode(factory.createStringLiteral(slug)),
                ),
              ],
            );
            parameterNodes.push(nestedFieldProperty);
          }

          return factory.createPropertySignature(
            undefined,
            slug,
            undefined,
            factory.createFunctionTypeNode(
              [
                factory.createTypeParameterDeclaration(
                  undefined,
                  typeArgumentIdentifiers.default,
                  undefined,
                  modelNode,
                ),
              ],
              [
                factory.createParameterDeclaration(
                  undefined,
                  undefined,
                  slug,
                  undefined,
                  factory.createUnionTypeNode(parameterNodes),
                ),
                sharedQueryOptionsParameter,
              ],
              promise
                ? factory.createTypeReferenceNode(identifiers.primitive.promise, [
                    factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
                  ])
                : factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
            ),
          );
        }),
      ),
    ]),
  );
};
