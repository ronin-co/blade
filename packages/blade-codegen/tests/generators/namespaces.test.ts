import { describe, expect, test } from 'bun:test';
import {
  blob,
  boolean,
  date,
  json,
  link,
  model,
  number,
  string,
} from 'blade-syntax/schema';

import { generateNamespaces } from '@/src/generators/namespaces';
import { printNodes } from '@/src/utils/print';

import type { PopulatedModel as Model } from 'blade-compiler';

describe('namespaces', () => {
  test('with a basic model', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        avatar: blob(),
        email: string({ required: true }),
        isActive: boolean(),
        lastActiveAt: date(),
        name: string(),
        rewardPoints: number({ defaultValue: 0, required: true }),
        settings: json({ defaultValue: {}, required: true }),
      },
    }) as unknown as Model;

    const typesResult = generateNamespaces([AccountModel]);

    expect(typesResult).toHaveLength(5);

    const typesResultStr = printNodes(typesResult);

    expect(typesResultStr).toMatchSnapshot();
  });

  test('with a link field', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        email: string({ required: true }),
      },
    }) as unknown as Model;

    const PostModel = model({
      slug: 'post',
      pluralSlug: 'posts',
      fields: {
        title: string(),
        author: link({ target: 'account' }),
      },
    }) as unknown as Model;

    const typesResult = generateNamespaces([AccountModel, PostModel]);

    expect(typesResult).toHaveLength(6);

    const typesResultStr = printNodes(typesResult);

    expect(typesResultStr).toMatchSnapshot();
  });

  test('with no models', () => {
    const typesResult = generateNamespaces([]);

    expect(typesResult).toHaveLength(4);

    const typesResultStr = printNodes(typesResult);

    expect(typesResultStr).toMatchSnapshot();
  });
});
