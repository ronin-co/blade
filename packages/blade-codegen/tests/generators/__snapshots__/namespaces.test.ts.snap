// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`namespaces with a basic model 1`] = `
"declare namespace AccountSyntax {
    type FieldSlug = "id" | "ronin.createdAt" | "ronin.createdBy" | "ronin.updatedAt" | "ronin.updatedBy" | "avatar" | "email" | "isActive" | "lastActiveAt" | "name" | "rewardPoints" | "settings";
    namespace Singular {
        type AfterQuery = Utils.AfterQuery<Account | null>;
        type AfterQueryPromise = Utils.AfterQueryPromise<Account | null>;
        type BeforeQuery = Utils.BeforeQuery<Account | null>;
        type BeforeQueryPromise = Utils.BeforeQueryPromise<Account | null>;
        type IncludingQuery = Utils.IncludingQuery<Account | null>;
        type IncludingQueryPromise = Utils.IncludingQueryPromise<Account | null>;
        type LimitedToQuery = Utils.LimitedToQuery<Account | null>;
        type LimitedToQueryPromise = Utils.LimitedToQueryPromise<Account | null>;
        type OrderedByQuery = Utils.OrderedByQuery<Account | null, AccountSyntax.FieldSlug>;
        type OrderedByQueryPromise = Utils.OrderedByQueryPromise<Account | null, AccountSyntax.FieldSlug>;
        type RootQueryCaller = Utils.RootQueryCaller<Account | null>;
        type RootQueryCallerPromise = Utils.RootQueryCallerPromise<Account | null>;
        type SelectingQuery = Utils.SelectingQuery<Account | null, AccountSyntax.FieldSlug>;
        type SelectingQueryPromise = Utils.SelectingQueryPromise<Account | null, AccountSyntax.FieldSlug>;
        type ToQuery = any;
        type ToQueryPromise = any;
        type UsingQuery = ReducedFunction & (<T = Account | null>(value: CombinedInstructions["using"]) => T);
        type UsingQueryPromise = ReducedFunction & (<T = Account | null>(value: CombinedInstructions["using"]) => Promise<T>);
        type WithQuery = Utils.WithQuery<Account | null> & {
            avatar: <T = Account | null>(avatar: Account["avatar"]) => T;
            email: <T = Account | null>(email: Account["email"]) => T;
            isActive: <T = Account | null>(isActive: Account["isActive"]) => T;
            lastActiveAt: <T = Account | null>(lastActiveAt: Account["lastActiveAt"]) => T;
            name: <T = Account | null>(name: Account["name"]) => T;
            rewardPoints: <T = Account | null>(rewardPoints: Account["rewardPoints"]) => T;
            settings: <T = Account | null>(settings: Account["settings"]) => T;
        };
        type WithQueryPromise = Utils.WithQueryPromise<Account | null> & {
            avatar: <T = Account | null>(avatar: Account["avatar"]) => Promise<T>;
            email: <T = Account | null>(email: Account["email"]) => Promise<T>;
            isActive: <T = Account | null>(isActive: Account["isActive"]) => Promise<T>;
            lastActiveAt: <T = Account | null>(lastActiveAt: Account["lastActiveAt"]) => Promise<T>;
            name: <T = Account | null>(name: Account["name"]) => Promise<T>;
            rewardPoints: <T = Account | null>(rewardPoints: Account["rewardPoints"]) => Promise<T>;
            settings: <T = Account | null>(settings: Account["settings"]) => Promise<T>;
        };
    }
    namespace Plural {
        type AfterQuery = Utils.AfterQuery<Accounts>;
        type AfterQueryPromise = Utils.AfterQueryPromise<Accounts>;
        type BeforeQuery = Utils.BeforeQuery<Accounts>;
        type BeforeQueryPromise = Utils.BeforeQueryPromise<Accounts>;
        type IncludingQuery = Utils.IncludingQuery<Accounts>;
        type IncludingQueryPromise = Utils.IncludingQueryPromise<Accounts>;
        type LimitedToQuery = Utils.LimitedToQuery<Accounts>;
        type LimitedToQueryPromise = Utils.LimitedToQueryPromise<Accounts>;
        type OrderedByQuery = Utils.OrderedByQuery<Accounts, AccountSyntax.FieldSlug>;
        type OrderedByQueryPromise = Utils.OrderedByQueryPromise<Accounts, AccountSyntax.FieldSlug>;
        type RootQueryCaller = Utils.RootQueryCaller<Accounts>;
        type RootQueryCallerPromise = Utils.RootQueryCallerPromise<Accounts>;
        type SelectingQuery = Utils.SelectingQuery<Accounts, AccountSyntax.FieldSlug>;
        type SelectingQueryPromise = Utils.SelectingQueryPromise<Accounts, AccountSyntax.FieldSlug>;
        type ToQuery = any;
        type ToQueryPromise = any;
        type UsingQuery = ReducedFunction & (<T = Accounts>(value: CombinedInstructions["using"]) => T);
        type UsingQueryPromise = ReducedFunction & (<T = Accounts>(value: CombinedInstructions["using"]) => Promise<T>);
        type WithQuery = Utils.WithQuery<Accounts> & {
            avatar: <T = Accounts>(avatar: Account["avatar"]) => T;
            email: <T = Accounts>(email: Account["email"]) => T;
            isActive: <T = Accounts>(isActive: Account["isActive"]) => T;
            lastActiveAt: <T = Accounts>(lastActiveAt: Account["lastActiveAt"]) => T;
            name: <T = Accounts>(name: Account["name"]) => T;
            rewardPoints: <T = Accounts>(rewardPoints: Account["rewardPoints"]) => T;
            settings: <T = Accounts>(settings: Account["settings"]) => T;
        };
        type WithQueryPromise = Utils.WithQueryPromise<Accounts> & {
            avatar: <T = Accounts>(avatar: Account["avatar"]) => Promise<T>;
            email: <T = Accounts>(email: Account["email"]) => Promise<T>;
            isActive: <T = Accounts>(isActive: Account["isActive"]) => Promise<T>;
            lastActiveAt: <T = Accounts>(lastActiveAt: Account["lastActiveAt"]) => Promise<T>;
            name: <T = Accounts>(name: Account["name"]) => Promise<T>;
            rewardPoints: <T = Accounts>(rewardPoints: Account["rewardPoints"]) => Promise<T>;
            settings: <T = Accounts>(settings: Account["settings"]) => Promise<T>;
        };
    }
}
"
`;

exports[`namespaces with a link field 1`] = `
"declare namespace AccountSyntax {
    type FieldSlug = "id" | "ronin.createdAt" | "ronin.createdBy" | "ronin.updatedAt" | "ronin.updatedBy" | "name" | "email";
    namespace Singular {
        type AfterQuery = Utils.AfterQuery<Account | null>;
        type AfterQueryPromise = Utils.AfterQueryPromise<Account | null>;
        type BeforeQuery = Utils.BeforeQuery<Account | null>;
        type BeforeQueryPromise = Utils.BeforeQueryPromise<Account | null>;
        type IncludingQuery = Utils.IncludingQuery<Account | null>;
        type IncludingQueryPromise = Utils.IncludingQueryPromise<Account | null>;
        type LimitedToQuery = Utils.LimitedToQuery<Account | null>;
        type LimitedToQueryPromise = Utils.LimitedToQueryPromise<Account | null>;
        type OrderedByQuery = Utils.OrderedByQuery<Account | null, AccountSyntax.FieldSlug>;
        type OrderedByQueryPromise = Utils.OrderedByQueryPromise<Account | null, AccountSyntax.FieldSlug>;
        type RootQueryCaller = Utils.RootQueryCaller<Account | null>;
        type RootQueryCallerPromise = Utils.RootQueryCallerPromise<Account | null>;
        type SelectingQuery = Utils.SelectingQuery<Account | null, AccountSyntax.FieldSlug>;
        type SelectingQueryPromise = Utils.SelectingQueryPromise<Account | null, AccountSyntax.FieldSlug>;
        type ToQuery = any;
        type ToQueryPromise = any;
        type UsingQuery = ReducedFunction & (<T = Account | null>(value: CombinedInstructions["using"]) => T);
        type UsingQueryPromise = ReducedFunction & (<T = Account | null>(value: CombinedInstructions["using"]) => Promise<T>);
        type WithQuery = Utils.WithQuery<Account | null> & {
            name: <T = Account | null>(name: Account["name"]) => T;
            email: <T = Account | null>(email: Account["email"]) => T;
        };
        type WithQueryPromise = Utils.WithQueryPromise<Account | null> & {
            name: <T = Account | null>(name: Account["name"]) => Promise<T>;
            email: <T = Account | null>(email: Account["email"]) => Promise<T>;
        };
    }
    namespace Plural {
        type AfterQuery = Utils.AfterQuery<Accounts>;
        type AfterQueryPromise = Utils.AfterQueryPromise<Accounts>;
        type BeforeQuery = Utils.BeforeQuery<Accounts>;
        type BeforeQueryPromise = Utils.BeforeQueryPromise<Accounts>;
        type IncludingQuery = Utils.IncludingQuery<Accounts>;
        type IncludingQueryPromise = Utils.IncludingQueryPromise<Accounts>;
        type LimitedToQuery = Utils.LimitedToQuery<Accounts>;
        type LimitedToQueryPromise = Utils.LimitedToQueryPromise<Accounts>;
        type OrderedByQuery = Utils.OrderedByQuery<Accounts, AccountSyntax.FieldSlug>;
        type OrderedByQueryPromise = Utils.OrderedByQueryPromise<Accounts, AccountSyntax.FieldSlug>;
        type RootQueryCaller = Utils.RootQueryCaller<Accounts>;
        type RootQueryCallerPromise = Utils.RootQueryCallerPromise<Accounts>;
        type SelectingQuery = Utils.SelectingQuery<Accounts, AccountSyntax.FieldSlug>;
        type SelectingQueryPromise = Utils.SelectingQueryPromise<Accounts, AccountSyntax.FieldSlug>;
        type ToQuery = any;
        type ToQueryPromise = any;
        type UsingQuery = ReducedFunction & (<T = Accounts>(value: CombinedInstructions["using"]) => T);
        type UsingQueryPromise = ReducedFunction & (<T = Accounts>(value: CombinedInstructions["using"]) => Promise<T>);
        type WithQuery = Utils.WithQuery<Accounts> & {
            name: <T = Accounts>(name: Account["name"]) => T;
            email: <T = Accounts>(email: Account["email"]) => T;
        };
        type WithQueryPromise = Utils.WithQueryPromise<Accounts> & {
            name: <T = Accounts>(name: Account["name"]) => Promise<T>;
            email: <T = Accounts>(email: Account["email"]) => Promise<T>;
        };
    }
}
declare namespace PostSyntax {
    type FieldSlug = "id" | "ronin.createdAt" | "ronin.createdBy" | "ronin.updatedAt" | "ronin.updatedBy" | "title" | "author";
    namespace Singular {
        type AfterQuery = Utils.AfterQuery<Post | null>;
        type AfterQueryPromise = Utils.AfterQueryPromise<Post | null>;
        type BeforeQuery = Utils.BeforeQuery<Post | null>;
        type BeforeQueryPromise = Utils.BeforeQueryPromise<Post | null>;
        type IncludingQuery = Utils.IncludingQuery<Post | null>;
        type IncludingQueryPromise = Utils.IncludingQueryPromise<Post | null>;
        type LimitedToQuery = Utils.LimitedToQuery<Post | null>;
        type LimitedToQueryPromise = Utils.LimitedToQueryPromise<Post | null>;
        type OrderedByQuery = Utils.OrderedByQuery<Post | null, PostSyntax.FieldSlug>;
        type OrderedByQueryPromise = Utils.OrderedByQueryPromise<Post | null, PostSyntax.FieldSlug>;
        type RootQueryCaller = Utils.RootQueryCaller<Post | null>;
        type RootQueryCallerPromise = Utils.RootQueryCallerPromise<Post | null>;
        type SelectingQuery = Utils.SelectingQuery<Post | null, PostSyntax.FieldSlug>;
        type SelectingQueryPromise = Utils.SelectingQueryPromise<Post | null, PostSyntax.FieldSlug>;
        type ToQuery = any;
        type ToQueryPromise = any;
        type UsingQuery = ReducedFunction & {
            <U extends Array<"author"> | "all">(fields: U): Post<U> | null;
            <T = Post | null>(fields: Array<"author"> | "all"): T;
        };
        type UsingQueryPromise = ReducedFunction & {
            <U extends Array<"author"> | "all">(fields: U): Promise<Post<U>> | null;
            <T = Post | null>(fields: Array<"author"> | "all"): Promise<T>;
        };
        type WithQuery = Utils.WithQuery<Post | null> & {
            title: <T = Post | null>(title: Post["title"]) => T;
            author: <T = Post | null>(author: Post["author"] | Partial<Post<[
                "author"
            ]>["author"]>) => T;
        };
        type WithQueryPromise = Utils.WithQueryPromise<Post | null> & {
            title: <T = Post | null>(title: Post["title"]) => Promise<T>;
            author: <T = Post | null>(author: Post["author"] | Partial<Post<[
                "author"
            ]>["author"]>) => Promise<T>;
        };
    }
    namespace Plural {
        type AfterQuery = Utils.AfterQuery<Posts>;
        type AfterQueryPromise = Utils.AfterQueryPromise<Posts>;
        type BeforeQuery = Utils.BeforeQuery<Posts>;
        type BeforeQueryPromise = Utils.BeforeQueryPromise<Posts>;
        type IncludingQuery = Utils.IncludingQuery<Posts>;
        type IncludingQueryPromise = Utils.IncludingQueryPromise<Posts>;
        type LimitedToQuery = Utils.LimitedToQuery<Posts>;
        type LimitedToQueryPromise = Utils.LimitedToQueryPromise<Posts>;
        type OrderedByQuery = Utils.OrderedByQuery<Posts, PostSyntax.FieldSlug>;
        type OrderedByQueryPromise = Utils.OrderedByQueryPromise<Posts, PostSyntax.FieldSlug>;
        type RootQueryCaller = Utils.RootQueryCaller<Posts>;
        type RootQueryCallerPromise = Utils.RootQueryCallerPromise<Posts>;
        type SelectingQuery = Utils.SelectingQuery<Posts, PostSyntax.FieldSlug>;
        type SelectingQueryPromise = Utils.SelectingQueryPromise<Posts, PostSyntax.FieldSlug>;
        type ToQuery = any;
        type ToQueryPromise = any;
        type UsingQuery = ReducedFunction & {
            <U extends Array<"author"> | "all">(fields: U): Post<U>;
            <T = Posts>(fields: Array<"author"> | "all"): T;
        };
        type UsingQueryPromise = ReducedFunction & {
            <U extends Array<"author"> | "all">(fields: U): Promise<Post<U>>;
            <T = Posts>(fields: Array<"author"> | "all"): Promise<T>;
        };
        type WithQuery = Utils.WithQuery<Posts> & {
            title: <T = Posts>(title: Post["title"]) => T;
            author: <T = Posts>(author: Post["author"] | Partial<Post<[
                "author"
            ]>["author"]>) => T;
        };
        type WithQueryPromise = Utils.WithQueryPromise<Posts> & {
            title: <T = Posts>(title: Post["title"]) => Promise<T>;
            author: <T = Posts>(author: Post["author"] | Partial<Post<[
                "author"
            ]>["author"]>) => Promise<T>;
        };
    }
}
"
`;

exports[`namespaces with no models 1`] = `""`;
